webrtc/RTCPeerConnection-ondatachannel.html

```diff
@@ -5,180 +5,388 @@
+  const pc2 = new RTCPeerConnection();
+  t.add_cleanup(() => pc1.close());
+  t.add_cleanup(() => pc2.close());
+
+  pc2.ondatachannel = t.step_func((event) => {
+    const dc = event.channel;
+    dc.onopen = t.step_func(() => {
+      assert_unreached('Open event should not fire');
     });
 
-    const localChannel = localPc.createDataChannel('test', {
-      id: 8,
-      ordered: false,
-      maxRetransmits: 1,
-      protocol: 'custom',
-      priority: 'high'
+    // This should prevent triggering the 'open' event
+    dc.close();
+
+    // Wait a bit to ensure the 'open' event does NOT fire
+    t.step_timeout(() => resolver.resolve(), 500);
+  });
+
+  pc1.createDataChannel('fire-me!');
+
+  exchangeIceCandidates(pc1, pc2);
+  await doSignalingHandshake(pc1, pc2);
+
+  await resolver;
+}, 'Open event should not be raised when closing the channel in the datachannel event');
+
+// Added this test as a result of the discussion in
+// https://github.com/w3c/webrtc-pc/pull/1851#discussion_r185976747
+promise_test(async (t) => {
+  const resolver = new Resolver();
+  const pc1 = new RTCPeerConnection();
+  const pc2 = new RTCPeerConnection();
+  t.add_cleanup(() => pc1.close());
+  t.add_cleanup(() => pc2.close());
+
+  pc2.ondatachannel = t.step_func((event) => {
+    const dc = event.channel;
+    dc.onopen = t.step_func((event) => {
+      resolver.resolve();
     });
 
-    assert_equals(localChannel.label, 'test');
-    assert_equals(localChannel.id, 8);
-    assert_equals(localChannel.ordered, false);
-    assert_equals(localChannel.maxRetransmits, 1);
-    assert_equals(localChannel.protocol, 'custom');
-    assert_equals(localChannel.priority, 'high');
-
-    remotePc.addEventListener('datachannel', onDataChannel);
-    exchangeIceCandidates(localPc, remotePc);
-    doSignalingHandshake(localPc, remotePc);
-  }, 'Data channel created on remote peer should match the same configuration as local peer');
-
-  /*
-    6.2.  RTCDataChannel
-      Dictionary RTCDataChannelInit Members
-        negotiated
-          The default value of false tells the user agent to announce the
-          channel in-band and instruct the other peer to dispatch a corresponding
-          RTCDataChannel object. If set to true, it is up to the application
-          to negotiate the channel and create a RTCDataChannel object with the
-          same id at the other peer.
-   */
-  async_test(t => {
-    const localPc = new RTCPeerConnection();
-    t.add_cleanup(() => localPc.close());
-    const remotePc = new RTCPeerConnection();
-
-    t.add_cleanup(() => remotePc.close());
-
-    const onDataChannel = t.unreached_func('datachannel event should not be fired');
-
-    localPc.createDataChannel('test', {
-      negotiated: true
+    // This should NOT prevent triggering the 'open' event since it enqueues at least two tasks
+    t.step_timeout(() => {
+      t.step_timeout(() => {
+        dc.close()
+      }, 1);
+    }, 1);
```

Comment:
Why 1 and not 0?

Reply:
For no particular reason. "[...] At least two tasks [...]".